cmd/juju/status/output_tabular.go

```diff
@@ -97,95 +97,98 @@ func FormatTabular(writer io.Writer, forceColor bool, value interface{}) error {
-			notes)
 
-		for un, u := range app.Units {
-			units[un] = u
-			if u.MeterStatus != nil {
-				metering = true
+		pUnit := func(name string, u unitStatus, level int) {
+			message := u.WorkloadStatusInfo.Message
+			agentDoing := agentDoing(u.JujuStatusInfo)
+			if agentDoing != "" {
+				message = fmt.Sprintf("(%s) %s", agentDoing, message)
 			}
+			if u.Leader {
+				name += "*"
+			}
+			w.Print(indent("", level*2, name))
+			w.PrintStatus(u.WorkloadStatusInfo.Current)
+			w.PrintStatus(u.JujuStatusInfo.Current)
+			p(
+				u.Machine,
+				u.PublicAddress,
+				strings.Join(u.OpenedPorts, ","),
+				message,
+			)
 		}
-	}
 
-	pUnit := func(name string, u unitStatus, level int) {
-		message := u.WorkloadStatusInfo.Message
-		agentDoing := agentDoing(u.JujuStatusInfo)
-		if agentDoing != "" {
-			message = fmt.Sprintf("(%s) %s", agentDoing, message)
-		}
-		if u.Leader {
-			name += "*"
-		}
-		w.Print(indent("", level*2, name))
-		w.PrintStatus(u.WorkloadStatusInfo.Current)
-		w.PrintStatus(u.JujuStatusInfo.Current)
-		p(
-			u.Machine,
-			u.PublicAddress,
-			strings.Join(u.OpenedPorts, ","),
-			message,
-		)
-	}
-
-	outputHeaders("Unit", "Workload", "Agent", "Machine", "Public address", "Ports", "Message")
-	for _, name := range naturalsort.Sort(stringKeysFromMap(units)) {
-		u := units[name]
-		pUnit(name, u, 0)
-		const indentationLevel = 1
-		recurseUnits(u, indentationLevel, pUnit)
-	}
-
-	if metering {
-		outputHeaders("Entity", "Meter status", "Message")
-		if fs.Model.MeterStatus != nil {
-			w.Print("model")
-			outputColor := fromMeterStatusColor(fs.Model.MeterStatus.Color)
-			w.PrintColor(outputColor, fs.Model.MeterStatus.Color)
-			w.PrintColor(outputColor, fs.Model.MeterStatus.Message)
-			w.Println()
-		}
+		outputHeaders("Unit", "Workload", "Agent", "Machine", "Public address", "Ports", "Message")
 		for _, name := range naturalsort.Sort(stringKeysFromMap(units)) {
 			u := units[name]
-			if u.MeterStatus != nil {
-				w.Print(name)
-				outputColor := fromMeterStatusColor(u.MeterStatus.Color)
-				w.PrintColor(outputColor, u.MeterStatus.Color)
-				w.PrintColor(outputColor, u.MeterStatus.Message)
+			pUnit(name, u, 0)
+			const indentationLevel = 1
+			recurseUnits(u, indentationLevel, pUnit)
+		}
+
+		if metering {
+			outputHeaders("Entity", "Meter status", "Message")
+			if fs.Model.MeterStatus != nil {
+				w.Print("model")
+				outputColor := fromMeterStatusColor(fs.Model.MeterStatus.Color)
+				w.PrintColor(outputColor, fs.Model.MeterStatus.Color)
+				w.PrintColor(outputColor, fs.Model.MeterStatus.Message)
 				w.Println()
 			}
+			for _, name := range naturalsort.Sort(stringKeysFromMap(units)) {
+				u := units[name]
+				if u.MeterStatus != nil {
+					w.Print(name)
+					outputColor := fromMeterStatusColor(u.MeterStatus.Color)
+					w.PrintColor(outputColor, u.MeterStatus.Color)
+					w.PrintColor(outputColor, u.MeterStatus.Message)
+					w.Println()
+				}
+			}
 		}
+		p()
+	}
+	if len(fs.Machines) > 0 {
```

Comment:
If if not doing the above suggestion, at least do it like this - an embedded printApplications() wrapped by an empty check

Reply:
done